portableappliancetest.py

```diff
@@ -0,0 +1,434 @@
+    fields = []
+    field_pack_format = {int: 'I'}
+
+    def __init__(self, endian='<'):
+        self.data = collections.OrderedDict()
+        self.build_format_string(endian=endian)
+
+    def fixup(self):
+        pass
+
+    def build_format_string(self, endian):
+        self.endian = endian
+        type_string = ''
+        for __, format_type, size in self.fields:
+            if format_type == int and size == 1:
+                type_string += 'B'
+            elif format_type == int and size == 2:
+                type_string += 'H'
+            elif format_type == int and size == 4:
+                type_string += 'L'
+            elif format_type == str:
+                type_string += str(size) + 's'
+            else:
+                type_string += self.field_pack_format[format_type]
+        self.format_string = self.endian + type_string
+        self.required_length = struct.calcsize(self.format_string)
+
+    def unpack(self, structure):
+        unpacked = list(struct.unpack(self.format_string, structure))
+        for name, format_type, __ in self.fields:
+            if format_type == str:
+                unpacked[0] = unpacked[0].replace(b'\x00', b'').rstrip()
+            self.data[name] = unpacked.pop(0)
+        return self
+
+    def headings(self):
+        return [name for name, format_type, size in self.fields]
+
+    def values(self):
+        return self.data.values()
+
+    def items_dict(self):
+        dictionary = '{'
+        dictionary += ', '.join(['%s:%s' % (key, value) for key, value in self.data.items()])
+        dictionary += '}'
+        return dictionary
+
+    def __len__(self):
+        return self.required_length
+
+    def __str__(self):
+        return str(self.data)
+
+# This sub-class for the SSS stream-format, most
+class SSS(Sdb):
+    def __init__(self):
+        super(SSS, self).__init__(endian='>')
+
+    def fixup(self):
+        pass
+
+    def unpack(self, structure):
+        unpacked = super(SSS, self).unpack(structure)
+        unpacked.fixup()
+        return self
+
+    def rescale(self, key):
+        self.data[key] = (10**-(self.data[key] >> 14)) * (self.data[key] & 0x3fff)
+
+    def passed(self, key='pass'):
+        self.data[key] = bool(self.data[key] == 1)
+
+class SSSRecordHeader(SSS):
+    fields = [('payload_length', int, 2),
+              ('nulls', int, 2),
+              ('checksum_header', int, 2)]
+
+    def checksum(self, payload):
+        # checksum is the sum value of all the bytes in the payload portion
+        self.data['checksum_payload'] = sum(payload) & 0xffff
+        match = (self.data['checksum_header'] == self.data['checksum_payload'])
+        self.data['checksum_match'] = match
+        return match
+
+class SSSVisualTest(SSS):
+    fields = [('id', str, 16),
+              ('hour', int, 1),
+              ('minute', int, 1),
+              ('day', int, 1),
+              ('month', int, 1),
+              ('year', int, 2),
```

Comment:
Replace with proper datetime

Reply:
This is a 1-to-1 mapping of the way that the tester reports records, this needs to remain in that format until post-processing